File: packages/mcp-server/src/tools/voice.ts
import { Tool } from "@modelcontextprotocol/sdk/types.js";
import { OneShot } from "@oneshot-agent/sdk";

export const voiceTool: Tool = {
  name: "oneshot_voice",
  description: "Make a phone call. Costs ~$0.25/minute. The call will follow the provided objective.",
  inputSchema: {
    type: "object",
    properties: {
      target_number: {
        type: "string",
        description: "Phone number to call (E.164 format, e.g., +14155551234)",
      },
      objective: {
        type: "string",
        description: "What the AI should accomplish on the call (min 10 characters)",
      },
      context: {
        type: "string",
        description: "Additional context about the call (optional)",
      },
      caller_persona: {
        type: "string",
        description: "Persona for the AI caller (optional)",
      },
      max_duration_minutes: {
        type: "number",
        description: "Maximum call duration in minutes (1-30, optional)",
      },
    },
    required: ["target_number", "objective"],
  },
};

export async function handleVoice(agent: OneShot, args: Record<string, unknown>) {
  return agent.voice({
    target_number: args.target_number as string,
    objective: args.objective as string,
    context: args.context as string | undefined,
    caller_persona: args.caller_persona as string | undefined,
    max_duration_minutes: args.max_duration_minutes as number | undefined,
  });
}

File: packages/langchain/langchain_oneshot/tools/voice.py
"""OneShot voice call tool."""

from __future__ import annotations

import json
from typing import Any, Type

from langchain_core.tools import BaseTool
from pydantic import BaseModel

from langchain_oneshot._types import VoiceInput
from oneshot import OneShotClient


class VoiceTool(BaseTool):
    """Make an autonomous phone call via OneShot. Paid tool.minute."""

    name: str = "oneshot_voice"
    description: str = (
        "Make a phone call with an AI agent. Provide an objective describing what "
        "the call should accomplish, and a target_number in E.164 format. Paid tool.min."
    )
    args_schema: Type[BaseModel] = VoiceInput
    handle_tool_error: bool = True

    client: OneShotClient

    def _run(self, **kwargs: Any) -> str:
        result = self.client.call_tool(
            "/v1/tools/voice/call", _strip_none(kwargs), timeout_sec=300
        )
        return json.dumps(result)

    async def _arun(self, **kwargs: Any) -> str:
        result = await self.client.acall_tool(
            "/v1/tools/voice/call", _strip_none(kwargs), timeout_sec=300
        )
        return json.dumps(result)


def _strip_none(d: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in d.items() if v is not None}
